### ISSUE TITLE
Automatically skip intro fluff for all videos in a playlist

### ISSUE THREAD

**Issue by imaverage** [iina/iina#3181]
**What you want IINA to do:**
I want an option such that when I am watching videos in a playlist, I want it to automatically detect and skip the intro parts of any new videos that play, based on what I've already seen in the previous videos. For example, if I'm watching season 1 of Spongebob, then I would only see the intro song once as I go through my playlist. The checkbox option could be called "Automatically detect and skip video intros based on previous videos in the playlist." with an optional offset timer (in case people want a sort of quick intermission at a duration of their own choosing).

Basic algorithm idea:
 * For each video, see how many starting frames are the same as for the previous video. Then automatically skip them. If you wanted to be efficient, you could leverage binary search (with right side being like halfway point of the video so it doesn't catch the same ending credit scenes etc) and then when it detects and answer, verify using an iterative approach.
 * Alternatively, you could check audio instead of video frames if that could be more performant. Or ideally, both.
 * If the above algorithm proves too costly performance wise, you could consider preprocessing it while a video is playing and saving that metadata onto the playlist file if the option is checked.

**What IINA does currently:**
It just plays every video in the playlist from the start.

**Why you think this should be added:**
It would allow the users of IINA to watch videos and not require interaction with the computer/media system every time a new video starts.

**Examples of other projects that have something similar:**
No one! But IINA is incredible, and more importantly, it's _known_ for being incredible. This would further maintain and enforce that idea. No other media player had video seek thumbnails when I decided to switch over to IINA a while ago. This would certainly be an incredible user case for many people.

**Comment by anohren** [comment]
> Examples of other projects that have something similar:
No one!

Well...not exactly. Have a look at Plex.

**Comment by CabotCake** [comment]
I believe VLC had this too. Users could enter in a time unit to skip at beginning of video and I believe at end of video to avoid credits.

**Comment by VonZubinski** [comment]
@CabotCake VLC had some Plugins to Skip Intros & Chapters

Netflix, Plex and others are mostly using audio fingerprinting to detect specific parts in movies/series.
BetterVRV and Anime Skip are using a community based approach.

I would suggest a simpel solution like [PotPlayer's approach](http://www.dvbsupport.net/images/screen/pp_skiptitle2.jpg) would be more than sufficient and a nice addition to IINA for binge watching TV shows, series and anime 😊

**Comment by imaverage** [comment]
Yes, a static approach like that to start off with would be great!

**Comment by ViiMorales** [comment]
Some video containers have chapter metadata that shows which section is the OPENING/ENDING, INTRO/OUTRO of that file. IINA could leverage on that first since is a low cost solution (performance-wise), then fallback on one of the other methods you guys are saying.

**Comment by VonZubinski** [comment]
@ViiMorales Check this Screenshot of the [approach PotPlayer](https://hg1.funnyjunk.com/comments/And+then+enter+the+ones+you+want+to+skip+as+_f90e718548a401f72fd8b47ddd7cddba.png) is using. It has a field for custom Keywords to look out for, so it shouldn't be a problem to skip the Intro, Opening, Prelude, Theme or whatever ¯\_(ツ)_/¯   

+1 for fallback methods

**Comment by VonZubinski** [comment]
@HarukaMa @alejx @lhc70000 @saagarjha We would greatly appreciate it, if someone could lable this as "Feature Request: Playback" 

Thanks to the team and keep up the great work 👍

**Comment by sentix001** [comment]
Would love this to happen

**Comment by Fernando-Laydner** [comment]
Quick question... Lets say the file is organized into chapters... Like youtube videos or even torrents sometimes (not always) are divided into chapters (VLC and other players can detect them no problem), lets say it was standard to call the intro chapter "Intro", would it be possible to add a button to skip chapters with "intro" or "outro" as their names? Like I know it would be insane to retro update every single video file to include this (DVD files should already have this), but would this be possible? I have seen many places saying it was impossible to be done even with croudsourcing, maybe if the subtitles providers marked things like intros, but this could also be an alternative, right?

**Comment by astyagun** [comment]
Not a full solution, but a step in the right direction, that works now: https://github.com/rui-ddc/skip-intro

**Comment by gitubejct761** [comment]
At the very least, a skip to the next/prior timestamp button could be added alongside some visual markers for the timestamp. See the implementation in MPV media player.
